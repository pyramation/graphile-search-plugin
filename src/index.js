export const PgSearchPlugin = (builder, { pgSearchPrefix = 'tsv' }) => {
  builder.hook('GraphQLInputObjectType:fields', function addConditionInputField(
    fields,
    build,
    context
  ) {
    const { inflection } = build;
    const {
      scope: { isPgCondition, pgIntrospection: table },
      fieldWithHooks
    } = context;
    if (!isPgCondition || !table || table.kind !== 'class') {
      return fields;
    }

    const tsvs = table.attributes.filter(
      (attr) => attr.type.name === 'tsvector'
    );
    if (!tsvs.length) return fields;

    return build.extend(
      fields,
      tsvs.reduce((m, v) => {
        const key = inflection.camelCase(`${pgSearchPrefix}_${v.name}`);
        m[key] = fieldWithHooks(
          key,
          {
            type: build.graphql.GraphQLString
          },
          {}
        );
        return m;
      }, {})
    );
  });
  builder.hook(
    'GraphQLObjectType:fields:field:args',
    function addSqlWhereClause(args, build, context) {
      const { pgSql: sql, inflection } = build;
      const {
        scope: {
          isPgFieldConnection,
          isPgFieldSimpleCollection,
          pgFieldIntrospection: procOrTable,
          pgFieldIntrospectionTable: tableIfProc
        },
        addArgDataGenerator
      } = context;
      const table = tableIfProc || procOrTable;
      if (
        (!isPgFieldConnection && !isPgFieldSimpleCollection) ||
        !table ||
        table.kind !== 'class'
      ) {
        return args;
      }

      const tsvs = table.attributes.filter(
        (attr) => attr.type.name === 'tsvector'
      );
      if (!tsvs.length) return args;

      for (const tsv of tsvs) {
        const conditionFieldName = inflection.camelCase(
          pgSearchPrefix + '_' + tsv.name
        );

        const conditionGenerator = (
          value,
          { queryBuilder, sql, sqlTableAlias }
        ) => {
          if (value == null) {
            return;
          }

          const tsquery = sql.fragment`websearch_to_tsquery('english', ${sql.value(
            value
          )})`;

          queryBuilder.orderBy(
            sql.fragment`ts_rank ( ${sqlTableAlias}.${sql.identifier(
              tsv.name
            )}, ${tsquery} )`,
            false,
            false
          );

          return sql.fragment`${sqlTableAlias}.${sql.identifier(
            tsv.name
          )} @@ ${tsquery}`;
        };

        addArgDataGenerator(function conditionSQLBuilder({ condition }) {
          if (!condition || !(conditionFieldName in condition)) {
            return {};
          }
          const { [conditionFieldName]: conditionValue } = condition;
          return {
            pgQuery: (queryBuilder) => {
              const sqlCondition = conditionGenerator(
                conditionValue,
                {
                  queryBuilder,
                  sql,
                  sqlTableAlias: queryBuilder.getTableAlias()
                },
                build
              );
              if (sqlCondition) {
                queryBuilder.where(sqlCondition);
              }
            }
          };
        });
      }

      return args;
    },
    [],
    // Make sure we're loaded before PgConnectionArgOrderBy, otherwise
    // ordering added by conditions will be overridden by the default
    // ordering.
    ['PgConnectionArgOrderBy'],
    []
  );
};

export default PgSearchPlugin;
