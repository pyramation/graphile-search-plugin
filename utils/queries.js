import { gql } from 'graphile-test';

export const GoalsSearchViaCondition = gql`
  query GoalsSearchViaCondition($search: String!) {
    goals(condition: { searchTsv: $search }) {
      nodes {
        id
        title
        description
      }
    }
  }
`;

export const GoalsSearchViaFilter = gql`
  query GoalsSearchViaFilter($search: String!) {
    goals(filter: { tsv: { matches: $search } }) {
      nodes {
        id
        title
        description
      }
    }
  }
`;

export const GoalsSearchViaCondition2 = gql`
  query GoalsSearchViaCondition($search: String!) {
    goals(condition: { searchStsv: $search }) {
      nodes {
        id
        title
        description
      }
    }
  }
`;

export const GoalsSearchViaFilter2 = gql`
  query GoalsSearchViaFilter($search: String!) {
    goals(filter: { stsv: { matches: $search } }) {
      nodes {
        id
        title
        description
      }
    }
  }
`;
