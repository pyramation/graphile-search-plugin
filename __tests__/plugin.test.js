import '../utils/env';
import { GraphQLTest, env, snapshot } from 'graphile-test';
import {
  GoalsSearchViaCondition,
  GoalsSearchViaFilter,
  GoalsSearchViaCondition2,
  GoalsSearchViaFilter2
} from '../utils/queries';
import { PgSearchPlugin } from '../src';
import PgSimpleInflector from 'graphile-simple-inflector';
import ConnectionFilterPlugin from 'postgraphile-plugin-connection-filter';
import FulltextFilterPlugin from '@pyramation/postgraphile-plugin-fulltext-filter';

const { SCHEMA } = env;

const getDbString = () =>
  `postgres://${env.PGUSER}:${env.PGPASSWORD}@${env.PGHOST}:${env.PGPORT}/${env.PGDATABASE}`;

const { setup, teardown, graphQL } = GraphQLTest(
  {
    appendPlugins: [
      PgSimpleInflector,
      ConnectionFilterPlugin,
      FulltextFilterPlugin,
      PgSearchPlugin
    ],
    schema: SCHEMA,
    graphqlRoute: '/graphql'
  },
  getDbString()
);

beforeAll(async () => {
  await setup();
});
afterAll(async () => {
  await teardown();
});

it('GoalsSearchViaFilter (order not relevant)', async () => {
  await graphQL(async (query) => {
    const data = await query(GoalsSearchViaFilter, {
      search: 'fowl'
    });
    expect(snapshot(data)).toMatchSnapshot();
  });
});

it('GoalsSearchViaCondition (order is relevant)', async () => {
  await graphQL(async (query) => {
    const data = await query(GoalsSearchViaCondition, {
      search: 'fowl'
    });
    expect(snapshot(data)).toMatchSnapshot();
  });
});

it('GoalsSearchViaFilter (order not relevant)', async () => {
  await graphQL(async (query) => {
    const data = await query(GoalsSearchViaFilter2, {
      search: 'fowl'
    });
    expect(snapshot(data)).toMatchSnapshot();
  });
});

it('GoalsSearchViaCondition (order is relevant)', async () => {
  await graphQL(async (query) => {
    const data = await query(GoalsSearchViaCondition2, {
      search: 'fowl'
    });
    expect(snapshot(data)).toMatchSnapshot();
  });
});
