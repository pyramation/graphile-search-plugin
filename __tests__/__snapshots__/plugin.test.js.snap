// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GoalsSearchViaCondition (order is relevant) 1`] = `
Object {
  "data": Object {
    "goals": Object {
      "nodes": Array [
        Object {
          "description": "Second in years female given. Us firmament. She'd kind there let moved thing evening saying set whales a fowl heaven.",
          "id": "[ID]",
          "title": "green fowl",
        },
        Object {
          "description": "Appear evening that gathered saying. Sea subdue so fill stars. Bring is man divided behold fish their. Also won't fowl.",
          "id": "[ID]",
          "title": "evenings",
        },
        Object {
          "description": "Heaven. Tree creeping was. Gathered living dominion us likeness first subdue fill. Fowl him moveth fly also the is created.",
          "id": "[ID]",
          "title": "heaven",
        },
      ],
    },
  },
}
`;

exports[`GoalsSearchViaCondition (order is relevant) 2`] = `
Object {
  "data": Object {
    "goals": Object {
      "nodes": Array [
        Object {
          "description": "Second in years female given. Us firmament. She'd kind there let moved thing evening saying set whales a fowl heaven.",
          "id": "[ID]",
          "title": "green fowl",
        },
      ],
    },
  },
}
`;

exports[`GoalsSearchViaFilter (order not relevant) 1`] = `
Object {
  "data": Object {
    "goals": Object {
      "nodes": Array [
        Object {
          "description": "Appear evening that gathered saying. Sea subdue so fill stars. Bring is man divided behold fish their. Also won't fowl.",
          "id": "[ID]",
          "title": "evenings",
        },
        Object {
          "description": "Heaven. Tree creeping was. Gathered living dominion us likeness first subdue fill. Fowl him moveth fly also the is created.",
          "id": "[ID]",
          "title": "heaven",
        },
        Object {
          "description": "Second in years female given. Us firmament. She'd kind there let moved thing evening saying set whales a fowl heaven.",
          "id": "[ID]",
          "title": "green fowl",
        },
      ],
    },
  },
}
`;

exports[`GoalsSearchViaFilter (order not relevant) 2`] = `
Object {
  "data": Object {
    "goals": Object {
      "nodes": Array [
        Object {
          "description": "Second in years female given. Us firmament. She'd kind there let moved thing evening saying set whales a fowl heaven.",
          "id": "[ID]",
          "title": "green fowl",
        },
      ],
    },
  },
}
`;
